### DISCUSSION TITLE
Spindle PWM level shifting

### DISCUSSION THREAD

**Discussion by avizienis** [terjeio/grblHAL#156] [Q&A] [answered]
Hi, I have a spindle driver which accepts 0-5V analog signal for rpm control. With AVR version of Grbl I just fed the pwm output to it and it worked wonderfully since the driver has an RC input. Now after migrating to arm, the max voltage achievable is 3.3V. The internet suggests a myriad of solutions but I want to avoid analog at all costs due to EMI sensitivity.
My current plan is to upscale the PWM output to 5V p-p with a LM393 or basically any logic IC accepting 3.3V levels. The big question is, should I do this conversion near grbl or at the driver? They are separated by about a meter of cable. 5V supplies are available at both ends.
Any tips appreciated, thanks!

**Comment by phil-barrett** [comment, marked as answer]
I would use the driver +5V  to power the level shifter. The driver will filter out the PWM and use the result as an analog control so if the shifter is powered from a lower voltage it will limit the top speed of your spindle (actual_+5V/5).  One customer was getting much lower spindle top speed because his +5 came from USB and it was fairly low (4.3V if I recall correctly).

There are lots of ways to level shift.  I like the mosfet approach for this as it doesn't lose much, if any, voltage level. [Here's an example.](https://electronics.stackexchange.com/questions/367052/replace-bss138-with-ao3400a-in-level-shifter-circuit) You will need to run +3.3V along with gnd and signal, though.

**Reply by avizienis** [reply]
Yes, good point about the different power supplies!
However I have doubts about these one transistor shifters - they have inherently asymmetrical output impedance for high and low phases. Hence different rise/fall times of the integrator and sequentially nonlinear transfer function. To combat that I imagine R in RC of the driver would have to be >>10k that's feeding the shifter. Or decrease the feeder to 1k or lower, which might load 5V line too much.
Or maybe these effects are irrelevant at the frequencies grblHAL is outputting PWM and I'm overthinking things :)

**Reply by terjeio** [reply]
I use a[ three transistor design](https://github.com/terjeio/CNC_Boosterpack#bonus-design-ii-bldc-motor-driver-board-33v-or-5v-pwm-to-0---10v-dc-conversion) with an [optocoupler](https://github.com/terjeio/grblHAL/discussions/130#discussioncomment-200902), can be configured for galvanic isolation.

**Comment by phil-barrett** [comment]
easy enough to simulate it in SPICE and see what it does.

FWIW, you can control the spindle PWM frequency with $33.
